### PR TITLE
Added EMP Ability

### PR DESCRIPTION
**Pull request by Flowaria**

### CODE REVIEW

File: ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/Handlers/EMPGunSightHandler.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+using System.Text;
+using System.Text.RegularExpressions;
+using UnityEngine;
+
+namespace EECustom.Customizations.EnemyAbilities.Abilities.EMP.Handlers
+{
+    public class EMPGunSightHandler : EMPHandlerBase
+    {
+        public GameObject _sightPicture;
+        const string _regex = "Sight_[0-9]*_[G-g]lass";
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/Handlers/EMPGunSightHandler.cs:13]: Protection definition is missing, inconsistent with rest of codebase.
```suggestion
        private const string _regex = @"Sight_[0-9]*_[G-g]lass";
```

Just as a note: I would personally inline this at L17

File: ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/Handlers/EMPPlayerHudHandler.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+using System.Collections.Generic;
+using System.Text;
+using UnityEngine;
+using LevelGeneration;
+
+namespace EECustom.Customizations.EnemyAbilities.Abilities.EMP.Handlers
+{
+    public class EMPPlayerHudHandler : EMPHandlerBase
+    {
+        readonly List<RectTransformComp> _targets = new();
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/Handlers/EMPPlayerHudHandler.cs:11]: Protection definition is missing, inconsistent with rest of codebase.

```suggestion
        private readonly List<RectTransformComp> _targets = new();
```

File: ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/Handlers/EMPSentryHandler.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+using System.Text;
+using UnityEngine;
+
+namespace EECustom.Customizations.EnemyAbilities.Abilities.EMP.Handlers
+{
+    public class EMPSentryHandler : EMPHandlerBase
+    {
+        static Color _offColor = new() { r = 0, g = 0, b = 0, a = 0 };
+        SentryGunInstance _sentry;
+        SentryGunInstance_ScannerVisuals_Plane _visuals;
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/Handlers/EMPSentryHandler.cs:13]: Protection definition is missing, inconsistent with rest of codebase.

```suggestion
        private static Color _offColor = new() { r = 0, g = 0, b = 0, a = 0 };
        private SentryGunInstance _sentry;
        private SentryGunInstance_ScannerVisuals_Plane _visuals;
```

File: ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/Inject/Inject_PlayerBackpack.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+using EECustom.Customizations.EnemyAbilities.Abilities.EMP.Handlers;
+using Gear;
+using HarmonyLib;
+using Player;
+
+namespace EECustom.Customizations.EnemyAbilities.Abilities.EMP.Inject
+{
+    internal class Inject_PlayerBackpack
+    {
+        public static void Setup()
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/Inject/Inject_PlayerBackpack.cs:13]: This is unused

File: ExtraEnemyCustomization/Patches/Handlers/FlyerStuckHandler.cs [hunk 5]
```diff
@@ -20,7 +20,8 @@ public FlyerStuckHandler(IntPtr ptr) : base(ptr)
         {
         }
 
-        private void Update()
+#pragma warning disable IDE0051 // Remove unused private members
```
Comment (Kasuromi) [inline ExtraEnemyCustomization/Patches/Handlers/FlyerStuckHandler.cs:23]: Just a note: you can make the methods public or internal, or you can keep this.

**Review by Kasuromi — CHANGES_REQUESTED** [review]
On first glance over the code, just some minor consistency issues and some notes.

File: ExtraEnemyCustomization/CustomSettings/CustomProjectileManager.cs [hunk 6]
```diff
@@ -52,7 +52,7 @@ public static void GenerateProjectile(CustomProjectile projInfo)
             }
 
             var newPrefab = ProjectileManager.SpawnProjectileType(projInfo.BaseProjectile, Vector3.zero, Quaternion.identity);
-            UnityEngine.Object.DontDestroyOnLoad(newPrefab);
+            //UnityEngine.Object.DontDestroyOnLoad(newPrefab);
```
[the remark this answers is not in the document]
Comment (Flowaria) [reply ExtraEnemyCustomization/CustomSettings/CustomProjectileManager.cs:55]: Yeah New code spawns Projectile Object related to ProjectileManager which should be consistent between scene

File: ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/EMPController.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        private IEMPHandler _handler = null;
+        private bool _hasHandler = false;
+        private float _duration;
+        private bool _setup = false;
+        private bool IsEMPActive => _duration > Clock.Time;
+
+        public EMPController(IntPtr ptr) : base(ptr)
+        {
+        }
```
[the remark this answers is not in the document]
Comment (Flowaria) [reply ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/EMPController.cs:18]: Is that avilable on latest version of BepInEx GTFO pack unhollower? I would like to remove this ctors from every handler if that's the case

File: ExtraEnemyCustomization/Patches/Handlers/FlyerStuckHandler.cs (hunk 5 above)
Comment (Flowaria) [reply ExtraEnemyCustomization/Patches/Handlers/FlyerStuckHandler.cs:23]: maybe making class itself internal would makes sense, I'll apply that after this PR has merged

File: ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/Inject/Inject_PlayerBackpack.cs (hunk 4 above)
Comment (Flowaria) [reply ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/Inject/Inject_PlayerBackpack.cs:13]: came out it was actually needed
https://github.com/GTFO-Modding/ExtraEnemyCustomization/pull/24/commits/66b6e6216f692c503fb87aa12cab37c630566a02

File: ExtraEnemyCustomization/CustomSettings/CustomProjectileManager.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Flowaria) [reply ExtraEnemyCustomization/CustomSettings/CustomProjectileManager.cs]: Actually came out SpawnProjectileType would destroy projectiles between level reload, reverting to old UnityEngine.Object.Instantiate

File: ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/EMPController.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Kasuromi) [reply ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/EMP/EMPController.cs:18]: It is included in it yeah

**Review by Kasuromi — APPROVED** [review]
Looks good to me, @Panthr75 should probably have a look as well

**Pull request merged**